### PR TITLE
(opera) Add unattended uninstallation support

### LINKED ISSUE #2249: (opera) Silent uninstall not working
### Chocolatey Version

2.0.0

### Chocolatey License

None

### Package Version

100.0.4815.21

### Current Behaviour

I am trying to uninstall opera silently, but it does not seem to work. I always pop ups the Uninstall window from opera.

### Expected Behaviour

Should uninstall opera silently without required user input

### Steps To Reproduce

After installing opera with:
choco install opera -y

I tried to uninstall it with the following command:
choco uninstall opera -y

### Environment

```markdown
- Operating System: Windows 11
- PowerShell Version: 5.1.22621.1778
```

### Chocolatey Log

```shell
https://gist.github.com/DasThomy/ddad550cae55a6f12f0eefd12ea68f8d
```

### Anything else?

Either I missunderstood the silently uninstall or something seems to be wrong with it.

### PR DESCRIPTION
**Pull request by brogers5**
## Description

This changeset modifies `opera`'s uninstallation process to close any running instances of Opera, then runs an AutoHotkey script, to work around Opera's decision to disable the `/silent` switch in its uninstaller.

## Motivation and Context

Fixes #2249.

This should enable the package to complete uninstallation in an automated fashion.

I've added a new dependency on `chocolatey-core.extension` v1.3.4+ and `autohotkey.portable` v2.0.0+, as well as a new package parameter (`/RemoveUserData`) to facilitate the process and ensure possibly desired functionality is exposed for user consumption.

The uninstaller's workflow may change slightly if existing user data is detected. After initially pressing the Uninstall button, a survey page may appear in the uninstaller for the user to optionally provide feedback as to why it's being uninstalled. The AutoHotkey script should detect when this happens, and will press the Uninstall button again to skip the survey and progress the uninstallation as expected.

>**Warning**
>The AutoHotkey script could potentially stall if an error is encountered during the uninstallation. As the uninstaller itself recommends closing Opera before trying again, `opera`'s uninstall script will also close all processes originating from the install location (if detected) prior to executing the uninstaller, in an effort to avoid this.
>
>If another error condition is encountered, the AutoHotkey script will continue pressing the Try Again button until the window closes (due to it either succeeding or closing by other means).

#### Dev

- OS Build: Windows 11 Pro v10.0.22621.0 (64-bit)
- PowerShell version: 7.3.6
- Chocolatey version: 2.2.2

#### Chocolatey Test Environment

- Vagrant Box Version: 3.0.0
- VM Provider: Virtualbox (although practically Hyper-V, since I manually converted the resulting virtual disk)
- OS Build: Windows Server 2019 Datacenter v10.0.17763.0 (64-bit)
- PowerShell version: 5.1.17763.3770
- Chocolatey version: 2.2.2

#### Uninstall without User Data

>**Important**
>Opera (both the package and software) is assumed to have never been installed previously.

1. Create a test package with the proposed changes from my dev environment (i.e. `choco pack`).
2. In the Chocolatey Testing Environment, install the test package version (i.e. `choco install opera --source="'.;https://community.chocolatey.org/api/v2/'"` [to prioritize local packages while sourcing dependencies from the Community Repository]).
3. Uninstall the test package version (i.e. `choco uninstall opera`).
4. Confirm that both the software and package's uninstallation completes without issues.

>**Note**
>As a post-uninstall step, this case will open a browser instance (either Google Chrome, Mozilla Firefox, or Microsoft Edge, in descending order, regardless of the user's browser preference) pointed to a page that redirects to <https://www.opera.com/client/uninstall>. As far as I know, this behavior cannot be suppressed. Closing the browser was deliberately avoided in the interest of preventing possibly unexpected closures.

#### Uninstall with User Data Creation and Preservation

1. Reinstall the test package version (i.e. `choco install opera --source="'.'"`).
2. Launch Opera after installation (it should show a splash screen upon first launch), then close it to trigger creation of user data.
3. Uninstall the test package version (i.e. `choco uninstall opera`).
4. Confirm that both the software and package's uninstallation completes without issues.
5. Reinstall the test package version (i.e. `choco install opera --source="'.'"`).
6. Launch Opera after installation, and confirm that no splash screen appears.

#### Uninstall while Opera is Running

1. Uninstall the test package version (i.e. `choco uninstall opera`).
2. Confirm that Opera is closed prior to the uninstaller launching.
3. Confirm that both the software and package's uninstallation completes without issues.

#### Uninstall with User Data Removal

1. Reinstall the test package version (i.e. `choco install opera --source="'.'"`).
2. Uninstall the test package version with the `/RemoveUserData` package parameter (i.e. `choco uninstall opera --package-parameters="'/RemoveUserData'"`).
3. Confirm that both the software and package's uninstallation completes without issues.
4. Reinstall the test package version (i.e. `choco install opera --source="'.'"`).
5. Launch Opera after installation, and confirm a splash screen appears upon first launch (which should signal no detected user data).

## Screenshot (if appropriate, usually isn't needed):

N/A

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Migrated package (a package has been migrated from another repository)

## Checklist:

- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] I have updated the package description and it is less than 4000 characters.
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)
- [x] The added/modified package passed install/uninstall in the chocolatey test environment.
- [x] The changes only affect a single package (not including meta package).

### CODE REVIEW

File: automatic/opera/opera.nuspec [hunk 1]
```diff
@@ -30,6 +30,10 @@ For example: `--params '"/NoAutostart /NoDesktopShortcut /NoTaskbarShortcut"'`
     <releaseNotes>https://blogs.opera.com/desktop/changelog-for-102/#b4880.40</releaseNotes>
     <tags>browser opera cross-platform internet admin</tags>
     <packageSourceUrl>https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/opera</packageSourceUrl>
+    <dependencies>
+      <dependency id="chocolatey-core.extension" version="1.3.4" />
```
Comment (AdmiringWorm) [inline automatic/opera/opera.nuspec:34]: As this is a new dependency, set it to the latest version available.

```suggestion
      <dependency id="chocolatey-core.extension" version="1.4.0" />
```

File: automatic/opera/tools/chocolateyUninstall.ps1 [hunk 2]
```diff
@@ -2,17 +2,32 @@
...
+      silentArgs     = "/uninstall"
       fileType       = 'EXE'
       validExitCodes = @(0)
       file           = $_.UninstallString.replace(' /uninstall', '').trim('"')
     }
+
+    $installLocation = Get-AppInstallLocation -AppNamePattern $softwareNamePattern
+    if ($null -ne $installLocation) {
+      Remove-Process -PathFilter "$([regex]::Escape($installLocation)).*" | Out-Null
+    }
```
Comment (AdmiringWorm) [inline automatic/opera/tools/chocolateyUninstall.ps1:21]: Stopping the process should generally happen in a `chocolateyBeforeModify.ps1` script.

Can this be moved to that file?

**Review by AdmiringWorm — CHANGES_REQUESTED** [review]

File: automatic/opera/tools/chocolateyUninstall.ps1 (hunk 2 above)
Comment (brogers5) [reply automatic/opera/tools/chocolateyUninstall.ps1:21]: I actually was considering putting it there, but opted against it when I tested the upgrade case. Turns out an upgrade can be done successfully without closing the current browser instance(s), and will take effect on the next boot. I was a little hesitant to make the package close it anyway when there's no apparent need for it to be done.

Did you still want me to move this over?

File: automatic/opera/opera.nuspec (hunk 1 above)
Comment (brogers5) [reply automatic/opera/opera.nuspec]: Addressed in https://github.com/chocolatey-community/chocolatey-packages/pull/2308/commits/1526a54c0318c9066e2483d15cf90fd9e57c2898.

File: automatic/opera/tools/chocolateyUninstall.ps1 (hunk 2 above)
Comment (AdmiringWorm) [reply automatic/opera/tools/chocolateyUninstall.ps1:21]: Hmm, considering it runs successfully but just isn't fully upgraded, I am honestly not sure about what would be the appropriate way here.

@pauby and @JPRuskin, what are your thoughts on this?

Comment (JPRuskin) [reply automatic/opera/tools/chocolateyUninstall.ps1:21]: Ah, I like that behaviour - I think that upgrading in the background is ideal, though perhaps with a prompt to the user that they'll need to restart to benefit? Would vote for this being fine.

**Review by AdmiringWorm — APPROVED** [review]
LGTM

**Comment by AdmiringWorm** [comment]
@brogers5 your changes have been merged. Thanks for your contribution 👍

**Pull request merged**